File: nodejs/base/client/lib/service-client.js
const { ResourceEndpointClient } = require("./resource-endpoint-client");

class ServiceClient {
    constructor(service, authProvider) {
        const endpointsMap = {};

        for (const resourceEndpoint of service.resources) {
            endpointsMap[resourceEndpoint.resourceType] =
                new ResourceEndpointClient(resourceEndpoint, authProvider, service.authType, service.authContext);
        }

        this.hasResourceEndpoint = (resourceType) => {
            if (typeof resourceType === "function" && resourceType.name) {
                resourceType = resourceType.name;
            }
            return endpointsMap[resourceType] !== undefined;
        };

        this.getResourceEndpointClient = (resourceType) => {
            if (typeof resourceType === "function" && resourceType.name) {
                resourceType = resourceType.name;
            }
            return endpointsMap[resourceType];
        };

        this.getAllResourceEndpointClients = () => {
            return Object.keys(endpointsMap).map(k => endpointsMap[k]);
        };
    }
}

module.exports = {
    ServiceClient
};

File: dotnet/base/Mcma.Client/ServiceClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Mcma.Core;

namespace Mcma.Client
{
    public class ServiceClient
    {
        public ServiceClient(Service service, IAuthProvider authProvider = null)
        {
            Data = service;

            ResourcesByType =
                service.Resources != null
                    ? service.Resources.ToDictionary(r => r.ResourceType, r => new ResourceEndpointClient(r, authProvider, service.AuthType, service.AuthContext))
                    : new Dictionary<string, ResourceEndpointClient>();
        }

        public Service Data { get; }

        private IDictionary<string, ResourceEndpointClient> ResourcesByType { get; }

        public IEnumerable<ResourceEndpointClient> Resources => ResourcesByType.Values;

        public bool HasResourceEndpointClient(string resourceType)
            => ResourcesByType.ContainsKey(resourceType);

        public bool HasResourceEndpointClient(Type resourceType)
            => HasResourceEndpointClient(resourceType.Name);

        public bool HasResourceEndpointClient<T>()
            => HasResourceEndpointClient(typeof(T));

        public ResourceEndpointClient GetResourceEndpointClient(string resourceType)
            => ResourcesByType.ContainsKey(resourceType) ? ResourcesByType[resourceType] : null;
            
        public ResourceEndpointClient GetResourceEndpointClient(Type resourceType)
            => GetResourceEndpointClient(resourceType.Name);
            
        public ResourceEndpointClient GetResourceEndpointClient<T>()
            => GetResourceEndpointClient(typeof(T));

        public ResourceEndpointClient[] GetAllResourceEndpointClients()
            => ResourcesByType.Values.ToArray();
    }
}